R/generic-spec.R

```diff
@@ -1,40 +1,65 @@
-as_generic <- function(x) {
-  if (is_generic(x) || is_external_generic(x)) {
-    x
-  } else if (inherits(x, "genericFunction")) {
+as_generic <- function(x, env = parent.frame()) {
+  if (is_generic(x) || is_external_generic(x) || is_S4_generic(x)) {
     x
   } else if (is.function(x)) {
-    as_S3_generic(x)
+    as_S3_generic(x, env)
   } else {
     msg <- sprintf("`generic` must be a function, not a %s", obj_desc(x))
     stop(msg, call. = FALSE)
   }
 }
 
-as_S3_generic <- function(x) {
+as_S3_generic <- function(x, env = parent.frame()) {
   use_method <- find_call(body(x), quote(UseMethod))
   if (!is.null(use_method)) {
-    return(S3_generic(x, as.character(use_method[[2]])))
+    name <- as.character(use_method[[2]])
+    pkg <- find_S3_package(x, name, env)
+    return(S3_generic(x, name, pkg))
   } else {
     name <- find_base_name(x)
     if (name %in% names(base_ops)) {
       return(base_ops[[name]])
```

Comment:
I guess you'd never externally register for a base op, but it probably still works anyways because I think the things stored in this are actually S7 generics?